tensorflow_probability/python/experimental/bijectors/highway_flow.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+import tensorflow as tf
+
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+tfb = tfp.bijectors
+tfe = tf.experimental
+
+'''
+This initial implementation is based on the PyTorch implementation at
+https://github.com/LucaAmbrogioni/CascadingFlow/blob/main/modules/networks.py
+'''
+
+
+class TriResNet(tfb.Bijector):
+
+    # set activation=False for last block
+    def __init__(self, width, activation=True, validate_args=False, name='tri_res_net'):
+        super(TriResNet, self).__init__(
+            validate_args=validate_args,
+            forward_min_event_ndims=0,
+            name=name)
+
+        self.width = width
+        self.W = tf.Variable(np.random.normal(0, 0.01, (self.width, self.width)), trainable=True, dtype=tf.float32)
+        self.b = tf.Variable(np.random.normal(0, 0.01, (self.width,)), trainable=True, dtype=tf.float32)  # bias
+
+        # positive diagonal for U
+        self.d = tf.Variable(np.random.normal(0, 0.01, (self.width,)), trainable=True, dtype=tf.float32)
+
+        # lambda before sigmoid
+        self.pre_l = tf.Variable(np.random.normal(0, 0.1, (1,)), trainable=True, dtype=tf.float32)
+
+        self.activation = activation
+
+        self.masku = tfe.numpy.tril(tf.ones((self.width, self.width)), -1)
+        self.maskl = tfe.numpy.triu(tf.ones((self.width, self.width)), 1)
+
+    def get_l(self):
+        return tf.nn.sigmoid(self.pre_l)
+
+    def get_L(self):
```

Comment:
My initial understanding of `TransformedVariable` was that we specify our variable before the bijector, and what we get is the transformed variable according to the chosen bijector. However, it seems like it works the other way around, i.e. the initial_value should be defined as if it was already transformed by the bijector. Is my understanding correct?

If this is correct, then one way to define the matrix U could be:
```
initial_value = np.triu(np.random.normal(0., 1., (5, 5)), 1) + np.diag(tf.math.softplus(np.random.normal(0, 1, (5,))))
fill_scale_tril_bijector = tfb.FillScaleTriL(diag_bijector=tfb.Softplus(), diag_shift=None)
transpose_bijector = tfp.bijectors.Transpose(rightmost_transposed_ndims=2)
chain_of_bijectors = tfp.bijectors.Chain([transpose_bijector, fill_scale_tril_bijector])

upper_diagonal_weights_matrix = tfp.util.TransformedVariable(initial_value=initial_value,
                                                             bijector=chain_of_bijectors,
                                                             trainable=True)
```
where we obtain `initial_value` by doing the whole transformation ourselves.

Following the same procedure, then the `residual_fraction` (lambda) could be defined as:

```
self.residual_fraction = tfp.util.TransformedVariable(
            initial_value=tf.math.sigmoid(np.random.normal(0, 0.1, (1,))),
            bijector=tfb.Sigmoid(), trainable=True)
```

Is this the correct way to use `TransformedVariable`? And should the definition of the `initial_value` be done differently?

Reply:
Good questions -- you're right that `initial_value` (somewhat counterintuitively) is the initial value *after* the transformation. Since `FillScaleTriL.inverse` ignores entries above the diagonal, you can define `initial_value` just as `np.random.uniform(0., 1.), (width, width))` and the value of the `TransformedVariable` will be triangular (here, using `uniform` instead of `normal` to avoid negative values).

Instead of the `Transpose` bijector, it might be more efficient to pass `transpose_a=True` into `matvec`, would that work? (Also, it looks like the `matvec` args are reversed -- the matrix should be first, then the vector.)

`residual_fraction` looks good, except it might be simpler to define `initial_value` as a `uniform` sample (or as a fixed value -- in our implementation, this is a parameter we expose to the user, with a default of 0.5). Also, `tf.Variable`s are trainable by default, so you don't need to pass `trainable=True`.